### ISSUE TITLE
Warn in Inspector when an asset surpasses performance standards

### ISSUE THREAD

**Issue by Beethoven-n** [Yellow-Dog-Man/Resonite-Issues#2263]
### Is your feature request related to a problem? Please describe.

This feature, if implemented, would make clear to the user that user-generated content needs to be optimized just as much as the backend of the game.

### Describe the solution you'd like

When you select a slot in the Inspector, the Inspector should warn you of various optimization failures, such as textures over an arbitrary limited size, polygon counts of a mesh over an arbitrary limited count, and some others worth discussing.

### Describe alternatives you've considered

It's possible that a seperate tool can be made for this purpose using ProtoFlux, but since it wouldn't necessarily be an official tool, it may become unmaintained, or pressure may mount against the maintainer of the tool to stop working on it.

### Additional Context

honestly this is one thing vrchat does that resonite should do. you're even allowed to upload avatars with a Very Poor rating, you just get alarm bells from all over the program that you shouldn't do this

### Requesters

@beethoven__

**Comment by Frooxius** [comment]
I don't really see how we could do this at the level of individual assets or even slots. That's usually not enough to determine the performance impact of something.

E.g. say someone has 8K texture on something. Is it bad? Is it good? It depends!
- Is it one of many 8K and 4K textures on avatar for stuff it doesn't need that much detail? Bad!
- Is it a single atlased texture that allows all materials for the avatar and its props to be shared? That's actually very good!

Things like this would be better handled with performance analysis tools, which will analyze given hierarchy and point out potentially problematic aspects of it - e.g. total amount of texture data, materials, geometry, skinned geometry and so on.

**Comment by BlueCyro** [comment]
Arbitrary content rating is not something I think would be very useful, and would go so far as to say that it would be detrimental socially.

Performance of a given asset depends on a lot of variables and the context in which its used. Statically marking X resolution or Y number of polys as "bad" or "good" is not actually indicative of the performance impact of that asset given the wide variety of contexts or other factors it might be used in.

For example, skyboxes are the one place where you maybe __do__ want a high resolution texture reaching upwards of 8k or 16k. If you mark those resolutions as "very bad", how do you determine what context its being used in? There are case-by-case applications where a single high-resolution texture is warranted, and you'll never get all of them.

Terrain meshes sometimes have to be high-poly as well. One could make the case that you should almost always compose your maps out of primitive colliders, but mesh colliders are sometimes the only solution for exceedingly big or complex map geometry that needs accurate collisions. If you mark high poly meshes as being "very bad" you again miss this context, and the myriad of others where high poly content is warranted.

Finally, you enter the social stigma. Making a "good vs bad" performance rating will undoubtedly create a social divide in our current community and breed unhealthy content creation habits to skirt around the technicalities of the way the rating is implemented. This is something I've seen, done, and heard about when a platform implements a goal post or constraint that the user must aim for or work within - we get situations where users will work around those limits by doing things that actually cause **more** performance issues than if they weren't hounded about it.

There is an example from another platform that I can think of right away: There is a platform that has a scripting language you can use to move objects around. Due to some issue, they limited how far you could move something in one call to the `SetPosition` function, so what did people do? They called `SetPosition` N number of times to get around the issue. This causes you to have to write your code oddly and end up using way more processing time than you need to.

The same can be said for these limits - people could upload like 50 different meshes that are all technically "good" in terms of poly count, but actually make up a super high poly mesh when joined together. Or use 20 different "good" textures to comprise a very big "bad" one while retaining a good rating.

These performance markers are not only easily bypassed in an endless myriad of ways, but also harmful to the cohesion of our community.

**Comment by BlueCyro** [comment]
TL;DR I'm not against performance analysis tooling, but we should not mark people as being "good" or "bad" based on an arbitrary set of standards.

**Comment by Beethoven-n** [comment]
oh the intent isn't to mark individuals as good or bad. it's purely to warn the user of an asset that they may be inconveniencing those on lower end hardware with some decisions they make. how would this intent be better conveyed?

**Comment by BlueCyro** [comment]
> oh the intent isn't to mark individuals as good or bad. it's purely to warn the user of an asset that they may be inconveniencing those on lower end hardware with some decisions they make. how would this intent be better conveyed?

Froox explained this above. There'd be generic statistics that you can use to determine what's good and bad on the object. But we'd likely never do a "good" and "bad" performance rating that other users would see when examining you.

**Comment by Beethoven-n** [comment]
can we at least get something to hint the user towards a standard they might want to follow?
i don't really know how to optimize assets on my own, and i get the feeling there's more than just me on that

**Comment by stawze** [comment]
This issue seems well covered by #233

**Comment by shiftyscales** [comment]
I agree- Frooxius had noted above- this issue would be better handled by performance analysis tooling which is covered by the request of #233.
